### ISSUE TITLE
Low frame rate is back

### ISSUE THREAD

**Issue by jzy-chitong56** [SMUnlimited/AMAI#472]
I'm not sure where it was caused, but the game became increasingly serious after 20 minutes of playing with 23 AI players. It is suspected that it was caused by the cancellation of all `ModuloInteger waiting` in JOB

Also, I don't understand why `MicroUnitsJob` and `SeedNewTownsForPlayer` needs to operate in two separate unit groups, it seems completely unnecessary

**Comment by Kingytrewq** [comment]
Isn't it because the game engine is too old? Although I rarely play 24melee maps, during a test of a 24 player AI melee, an error window popped up after exiting the game. Low frame rates are common in later games, usually causing an increase in CPU frequency, and the number of cores is still 4, so this is still related to the game engine, not AI. At least I think so. In 3.4.1, there was no crash that occurred in the game, at least so far

**Comment by jzy-chitong56** [comment]
> Isn't it because the game engine is too old?
Not at all

**Comment by Kingytrewq** [comment]
Yes, this is the reason for the underlying code, or I found that once the map crashes, I need to reinstall amai. This may be due to the code being damaged in the map or some other reason. What do you think?

**Comment by jzy-chitong56** [comment]
The probability of crashing during the first installation is indeed very low, but it always crashes when running the flagship afterwards. If the map is damaged, it is theoretically impossible to enter the game. If it is a problem with the game itself, you can try other versions, such as 3.5X, which will be much smoother

**Comment by Kingytrewq** [comment]
The same problem, in fact, no matter what kind of updates and optimizations are made to Amai, there is always a certain probability of crashing. I found that this depends on the FPS of the game. In the recently updated PTR version, I got FPS below 200, which is obviously the underlying reason for the crash

**Comment by Kingytrewq** [comment]
I can understand that adding sleep can prevent crashes from occurring, but it cannot completely solve them, so crashes still have a certain probability of occurring, but it is not related to the code of the game itself, but rather a probability issue

**Comment by jzy-chitong56** [comment]
` call TQAddJob(4 + 1 * sleep_multiplier, TOWN_TRACK, 0)`
`call TQAddJob(1, ARMY_TRACK, 0)`

`ARMY_TRACK` runs two complete loops almost every 15 seconds, and the disabling of other `TQSLEEP` may be the reason for the PPT getting stuck

**Comment by jzy-chitong56** [comment]
The core issue originated from army tracking, and my version fixed this problem

- When calculating , some code are not set correctly  -- This resulted in abnormal army  assessment

like  here , 
 `exitwhen i != lastArmyNum and ModuloInteger(i,30) == 0`

You exited the loop, and lastArmyNum and lastFreeNum were correctly set to the values that need to be connected in the next loop

```
  if i >= army_num then
    set army_num = first_free
    set mode = 1
    set lastArmyNum = 0
    set lastFreeNum = 0
  else
    set lastArmyNum = army_num
    set lastFreeNum = first_free
    call TQAddJob(1, ARMY_TRACK, 0)
  endif
```

There are two questions here

- At this point, there are some empty data between i and first_free, but army_num is not tightly compressed, causing errors in checking army_num for other functions during this interval -- It doesn't matter
- in `ArmyTrackJob` ， run `call GroupClear(in_army_group)` ,  At this point, the loop has not been fully completed yet. As I mentioned earlier, it exited in the middle, which resulted in the entire in'army_group being recalculated in the next loop, and ExistingArmies being lost  --- This is the fundamental reason for the lag

In my modification, army_num was compressed and in'army_group was ensured to be cleared only when the loop was completed
https://github.com/jzy-chitong56/AMAI-CN/blob/2.6.x/Jobs/ARMY_TRACK.eai
https://github.com/jzy-chitong56/AMAI-CN/blob/2.6.x/Jobs/TOWN_TRACK.eai

```
if i >= army_num then
    set army_num = first_free
    set mode = 1
    set lastArmyNum = 0
    set lastFreeNum = 0
  else
    set lastArmyNum = i
    set lastFreeNum = first_free
    if i != first_free then
      loop
        exitwhen i >= army_num
        call CopyArmy(i, first_free)
        set first_free = first_free + 1
        set i = i + 1
      endloop
      set lastArmyNum = lastFreeNum
      set army_num = first_free
    endif
    call TQAddJob(1, ARMY_TRACK, 0)
  endif
```

```
function ArmyTrackJob takes nothing returns nothing
  //local unit alliance = null
  call DisplayToAllJobDebug("ARMY_TRACK JOB START")
  if mode == 0 then
    call TrackExistingArmies()
  endif
  if mode == 1 then
    call SeedNewArmies()
  endif
  if mode == 2 then
    if town_num > 0 then
      call UpdateTownThreat()
    else
      set mode = 3
      set lastFreeNum = 0
    endif
  endif
  if mode == 3 then
    set mode = 0
    set tmpmainarmy = -1
    call GroupClear(in_army_group)
    call TQAddJob(4 + 1 * sleep_multiplier, TOWN_TRACK, 0)
  endif
endfunction

```
